ScoutSuite/providers/azure/provider.py

```diff
@@ -35,39 +39,55 @@ def __init__(self, project_id=None, organization_id=None,
         """
-        Implement authentication for the GCP provider
-        Refer to https://google-auth.readthedocs.io/en/stable/reference/google.auth.html.
+        Implements authentication for the Azure provider using azure-cli.
+        Refer to https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python.
 
         :return:
         """
 
         try:
-
-            # TODO this is temporary
-            import os
-            # Azure subscription
-            SUBSCRIPTION_ID = os.environ['SUBSCRIPTION_ID']
-            # Tenant ID for your Azure Subscription
-            TENANT_ID = os.environ['TENANT_ID']
-            # Your Service Principal App ID
-            CLIENT = os.environ['CLIENT']
-            # Your Service Principal Password
-            KEY = os.environ['KEY']
-
-            self.aws_account_id = TENANT_ID # TODO this is for AWS
-
-            credentials = ServicePrincipalCredentials(
-                client_id=CLIENT,
-                secret=KEY,
-                tenant=TENANT_ID
-            )
-
-            self.credentials = AzureCredentials(credentials, SUBSCRIPTION_ID)
-
-            return True
-
+            if azure_cli:
+                cli_credentials, self.aws_account_id = get_azure_cli_credentials()  # TODO: Remove aws_account_id
+                self.credentials = AzureCredentials(cli_credentials, self.aws_account_id)
+                print(cli_credentials)
+                return True
+            elif azure_msi:
+                # TODO test
+                credentials = MSIAuthentication()
+                self.aws_account_id = "aaa"
+                self.credentials = AzureCredentials(credentials, self.aws_account_id)
+                return True
+            elif azure_file_auth:
+                # TODO
+                return True
+            elif azure_service_principal:
+                subscription_id = input("Subscription ID: ")
+                tenant_id = input("Tenant ID: ")
+                client_id = input("Client ID: ")
+                client_secret = getpass("Client secret: ")
+
+                self.aws_account_id = tenant_id  # TODO this is for AWS
+
+                credentials = ServicePrincipalCredentials(
+                    client_id=client_id,
+                    secret=client_secret,
+                    tenant=tenant_id
+                )
+
+                self.credentials = AzureCredentials(credentials, subscription_id)
+
+                return True
+            elif azure_user_credentials:
+                username = input("Username: ")
+                password = getpass("Password: ")
+
+                credentials = UserPassCredentials(username, password)
+                self.aws_account_id = ""  # TODO this is for AWS
+                self.credentials = AzureCredentials(credentials, self.aws_account_id)
+                return True
```

Comment:
I suggest we break down that method into multiple smaller methods. For example, `azure_cli_authentication`, `azure_msi_authentication`, `azure_file_auth`, etc. That would increase readability and 
 
It could be interesting to eventually implement a strategy pattern or something like that instead of that if/elif/elif/...

Reply:
Good idea, I'll do it!

Reply:
Hi!

If you want, I could check this out while I refactor the cli parser, as this pattern would also be useful for the other provider. For now, at least it's in a working state and I'll clean it up for good on the next PR!

Reply:
> I could check this out while I refactor the cli parser, as this pattern would also be useful for the other provider.

Agreed, the CLI implementation can/should be improved as currently it's just a soup of parameters. The parameters should have a hierarchy so that it's clear to which provider they apply.

Reply:
Fair enough 👍